File: crates/glyph-infer/src/postprocess.rs
//! Postprocessing: thresholded edge probabilities -> union-find connected
//! components -> contour groups representing character segments.

use std::collections::HashMap;

use glyph_core::GraphSample;

struct UnionFind {
    parent: Vec<u32>,
}

impl UnionFind {
    fn new(n: usize) -> Self {
        Self {
            parent: (0..n as u32).collect(),
        }
    }

    fn find(&mut self, x: u32) -> u32 {
        let mut root = x;
        while self.parent[root as usize] != root {
            root = self.parent[root as usize];
        }
        let mut cur = x;
        while self.parent[cur as usize] != root {
            let next = self.parent[cur as usize];
            self.parent[cur as usize] = root;
            cur = next;
        }
        root
    }

    fn union(&mut self, a: u32, b: u32) {
        let (ra, rb) = (self.find(a), self.find(b));
        if ra != rb {
            self.parent[ra as usize] = rb;
        }
    }
}

/// Groups contours into character segments.
///
/// Cross-contour edges are aggregated by max probability per contour pair.
/// Contours whose pair score clears `threshold` are merged with union-find.
/// Each contour is an atomic unit (intra-contour connectivity is implicit).
///
/// Returns contour-id groups sorted by leftmost node for stable output.
pub fn group_characters(g: &GraphSample, probs: &[f32], threshold: f32) -> Vec<Vec<u32>> {
    let n = g.num_nodes as usize;
    if n == 0 {
        return Vec::new();
    }
    let e = g.num_edges();
    let num_contours = g
        .node_contour_ids
        .iter()
        .copied()
        .max()
        .map(|m| m + 1)
        .unwrap_or(0) as usize;
    let (src, dst) = g.edge_index.split_at(e);

    let mut pair_prob: HashMap<(u32, u32), f32> = HashMap::new();
    for i in 0..e {
        let ca = g.node_contour_ids[src[i] as usize];
        let cb = g.node_contour_ids[dst[i] as usize];
        if ca == cb {
            continue;
        }
        let key = if ca < cb { (ca, cb) } else { (cb, ca) };
        let entry = pair_prob.entry(key).or_insert(0.0);
        if probs[i] > *entry {
            *entry = probs[i];
        }
    }

    let mut uf = UnionFind::new(num_contours);
    for (&(a, b), &prob) in &pair_prob {
        if prob >= threshold {
            uf.union(a, b);
        }
    }

    let mut groups: HashMap<u32, Vec<u32>> = HashMap::new();
    for contour in 0..num_contours as u32 {
        groups.entry(uf.find(contour)).or_default().push(contour);
    }

    let node_dim = g.node_dim as usize;
    let contour_min_x = |cids: &[u32]| -> f32 {
        let mut min_x = f32::INFINITY;
        for node in 0..n {
            if cids.contains(&g.node_contour_ids[node]) {
                min_x = min_x.min(g.node_features[node * node_dim]);
            }
        }
        min_x
    };
    let mut out: Vec<Vec<u32>> = groups
        .into_values()
        .map(|mut v| {
            v.sort_unstable();
            v
        })
        .collect();
    out.sort_by(|a, b| contour_min_x(a).total_cmp(&contour_min_x(b)));
    out
}

/// Ground-truth contour groups derived from stored char ids (for eval).
pub fn ground_truth_groups(g: &GraphSample) -> Vec<Vec<u32>> {
    let mut by_char: HashMap<u32, Vec<u32>> = HashMap::new();
    let n = g.num_nodes as usize;
    for node in 0..n {
        let cid = g.node_contour_ids[node];
        let ch = g.node_char_ids[node];
        let v = by_char.entry(ch).or_default();
        if !v.contains(&cid) {
            v.push(cid);
        }
    }
    let mut out: Vec<Vec<u32>> = by_char
        .into_values()
        .map(|mut v| {
            v.sort_unstable();
            v
        })
        .collect();
    out.sort();
    out
}

/// Set equality of groupings, ignoring order.
pub fn groups_equal(a: &[Vec<u32>], b: &[Vec<u32>]) -> bool {
    let mut sa: Vec<Vec<u32>> = a.to_vec();
    let mut sb: Vec<Vec<u32>> = b.to_vec();
    sa.sort();
    sb.sort();
    sa == sb
}

File: training/glyph_gnn/postprocess.py
"""Union-find postprocessing mirroring ``glyph-infer/src/postprocess.rs``."""

from __future__ import annotations

from collections import defaultdict


def group_characters(
    num_nodes: int,
    node_contour_ids: list[int],
    edge_src: list[int],
    edge_dst: list[int],
    probs: list[float],
    threshold: float,
) -> list[list[int]]:
    """Groups contours into character segments from thresholded edge probabilities."""
    parent = list(range(num_nodes))

    def find(x: int) -> int:
        root = x
        while parent[root] != root:
            root = parent[root]
        while parent[x] != root:
            nxt = parent[x]
            parent[x] = root
            x = nxt
        return root

    def union(a: int, b: int) -> None:
        ra, rb = find(a), find(b)
        if ra != rb:
            parent[ra] = rb

    for src, dst, prob in zip(edge_src, edge_dst, probs, strict=True):
        if prob >= threshold:
            union(src, dst)

    votes: dict[tuple[int, int], int] = defaultdict(int)
    for node in range(num_nodes):
        comp = find(node)
        contour = node_contour_ids[node]
        votes[(contour, comp)] += 1

    contour_comp: dict[int, tuple[int, int]] = {}
    for (contour, comp), count in votes.items():
        best_comp, best_count = contour_comp.get(contour, (comp, 0))
        if count > best_count:
            contour_comp[contour] = (comp, count)

    groups_map: dict[int, list[int]] = defaultdict(list)
    for contour, (comp, _) in contour_comp.items():
        groups_map[comp].append(contour)

    return [sorted(contours) for contours in groups_map.values()]


def sort_groups_by_x(groups: list[list[int]], node_x: list[float], node_contour_ids: list[int]) -> list[list[int]]:
    """Orders groups left-to-right using the leftmost node in each group."""

    def contour_min_x(cids: list[int]) -> float:
        min_x = float("inf")
        cid_set = set(cids)
        for node, contour in enumerate(node_contour_ids):
            if contour in cid_set:
                min_x = min(min_x, node_x[node])
        return min_x

    return sorted(groups, key=contour_min_x)
